Adding MapStore Offloading content. (#508)
hazelcast/hz-docs: docs/modules/data-structures/pages/working-with-external-data.adoc

TIP: link:https://github.com/hazelcast/hazelcast-code-samples/tree/master/distributed-map/mapstore/src/main/java[See more MapStore/MapLoader code samples].

== Offloading MapStore and MapLoader Operations

> To improve cluster throughput, all MapStore operations do not block partition threads.

Reviewer: ```suggestion
To improve cluster throughput, MapStore/MapLoader api calls are offloaded. This way partition threads are not blocked by them.
```

Each partition in a member is managed by a single thread, ensuring that the write operations to a given partition are handled
one at a time in first-in-first-out order. However, MapStore operations may take a long time to connect to the external data source and complete, blocking the affected partitions. 
To avoid blocking partition threads, MapStore operations that communicate with the external data source are handled asynchronously, using an <<configuring-the-offload-executor, executor>>.

=== Limitations

* The `init()` and `destroy()` methods of the `MapLoaderLifecycleSupport` interface cannot be offloaded. These methods are triggered only once in the whole lifecycle of a map.

> * MapStore methods `loadAll()` and `loadAllKeys()` cannot be offloaded.

Reviewer: >MapStore methods `loadAll()` and `loadAllKeys()` cannot be offloaded.

These methods are also offloaded independent of the work done in 5.2. In 5.1, they are also working on offloaded manner.

* Offloading cannot be used for xref:tiered-storage:overview.adoc[tiered storage], since you cannot enable both MapStore and tiered storage for a map.

=== Blocking Partition Threads

If you want to block partition threads until MapStore operations are complete, 

> you can configure the `offload` setting for the MapStore. It is enabled by default (`true`) and can be configured for each map in your cluster.

Reviewer: >If you want to block partition threads until MapStore operations are complete, 
you can configure the `offload` setting for the MapStore. It is enabled by default (`true`) and can be configured for each map in your cluster.

To restore the behavior in previous releases than 5.2, you can set `offload` setting for the `MapStoreConfig` to `false` which is `true` by default. This way partition threads will be blocked during MapStore/MapLoader api calls.

Author: We try to avoid mentioning the product release numbers, since the documentation is versioned.
Instead, I can simply give the idea of "set to "false" to block".

Reviewer: My point is we should emphasis that the point of making it false is only bringing old behavior back. Giving version number is not necessary.

TIP: link:https://github.com/hazelcast/hazelcast-code-samples/tree/master/distributed-map/mapstore/src/main/java[See more MapStore/MapLoader code samples].

> == Offloading MapStore and MapLoader Operations

Reviewer: @Serdaro you can have up-to-date info for the design here: https://github.com/hazelcast/hazelcast/blob/master/docs/design/map/map-store-operation-offloading.md

Reviewer: @Serdaro also worth to mention the name of the executor for the offloaded calls and how oen can configure it: https://github.com/hazelcast/hazelcast/blob/master/docs/design/map/map-store-operation-offloading.md#configuring-offloadable-executor

> Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block each other.

Reviewer: ```suggestion
Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block partition threads.
```

> Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block each other.

Reviewer 2: @ahmetmircik WDYT of this?

> Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block each other.

Reviewer: >To improve cluster throughput, most MapStore operations do not block partition threads.

Not most but all MapStore operations are async by default. Only `MapLoaderLifecycleSupport` is sync but it runs only one time during the creation of IMap.

> Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block each other.

Reviewer 2: > > MapStore operations that communicate with the external data source are handled asynchronously, using an entry processor.
> 
> I didn't get the relation between EP and async map-store calls. Async map-store calls are done independently, EP is not required.

It's an executor instead?

> Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block each other.

Reviewer 2: > > To improve cluster throughput, most MapStore operations do not block partition threads.
> 
> Not most but all MapStore operations are async by default. Only `MapLoaderLifecycleSupport` is sync but it runs only one time during the creation of IMap.

Since `MapLoaderLifecycleSupport` is sync, isn't it incorrect to say all?

> Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block each other.

Reviewer: >Since MapLoaderLifecycleSupport is sync, isn't it incorrect to say all?

No, it is a separate interface and implementing it is not required, it is up-to-user. If user is implemented it, it is blocking.

> Offloading `MapStore` and `MapLoader` operations allow these operations to work asynchronously, meaning they do not block each other.

Reviewer 2: > > It's an executor instead?
> 
> It is an executor behind the scenes, see this please: [#508 (comment)](https://github.com/hazelcast/hz-docs/pull/508#discussion_r951240867)

Got it. I think we should mention it here in that case.

> This provides faster executions of these operations, resulting in higher throughput of the whole cluster.

Reviewer: ```suggestion
This provides faster executions of other operations in cluster, since partition threads are not blocked by MapStore/MapLoader calls, resulting in higher throughput of the whole cluster.
```

For this, you can use the `offload` configuration element as shown below. It is enabled by default (`true`). You can configure the offloading for each map in your cluster.

[tabs] 
==== 
XML:: 
+ 
-- 
[source,xml]
----
<hazelcast>
    <map name="default">

>         <map-store enabled="true" initial-mode="LAZY">

Reviewer: ```suggestion
        <map-store enabled="true">
```
Having LAZY initial mode is not required.

            <offload>true</offload>
            ...
        </map-store>
    </map>
</hazelcast>
----
--

YAML::
+
[source,yaml]
----
hazelcast:
  map:
    default:
      map-store:
        enabled: true
        initial-mode: LAZY
        offload: true
        ...
----

Java::
+
[source,java]
----
Config config = new Config();
MapConfig mapConfig = config.getMapConfig();
MapStoreConfig mapStoreConfig = mapConfig.getMapStoreConfig();

> mapStoreConfig.setOffload("true");

Reviewer: ```suggestion
mapStoreConfig.setOffload(true);
```

...
----
====

> NOTE: Offloading cannot be used for xref:tiered-storage:overview.adoc[Tiered Storage], and for `MapLoaderLifecycleSupport`, since it is a one time operation per member along the lifecycle of a map.

Reviewer 2: I also see this in the design doc: `No changes in cluster wide map bulk loading mechanism.` Does this just mean the `loadAllKeys()` method?

> NOTE: Offloading cannot be used for xref:tiered-storage:overview.adoc[Tiered Storage], and for `MapLoaderLifecycleSupport`, since it is a one time operation per member along the lifecycle of a map.

Reviewer 2: I think we should move any exceptions up to the bulleted list that I suggested above.

> NOTE: Offloading cannot be used for xref:tiered-storage:overview.adoc[Tiered Storage], and for `MapLoaderLifecycleSupport`, since it is a one time operation per member along the lifecycle of a map.

Reviewer: @JakeSCahill currently you cannot have both map-store and tiered-storage enabled for a map. This is by design.

> NOTE: Offloading cannot be used for xref:tiered-storage:overview.adoc[Tiered Storage], and for `MapLoaderLifecycleSupport`, since it is a one time operation per member along the lifecycle of a map.

Reviewer 2: So maybe it's worth having a 'Limitations' section higher up this page? So users are aware of what doesn't work with MapStore?

> NOTE: Offloading cannot be used for xref:tiered-storage:overview.adoc[Tiered Storage], and for `MapLoaderLifecycleSupport`, since it is a one time operation per member along the lifecycle of a map.

Reviewer 2: This location doesn't seem like the place to mention a limitation that affects all MapStores.

> NOTE: Offloading cannot be used for xref:tiered-storage:overview.adoc[Tiered Storage], and for `MapLoaderLifecycleSupport`, since it is a one time operation per member along the lifecycle of a map.

Reviewer: makes sense.

Config config = new Config();
MapConfig mapConfig = config.getMapConfig();
MapStoreConfig mapStoreConfig = mapConfig.getMapStoreConfig();
mapStoreConfig.setOffload(true);
...
----
====

=== Configuring the Offload Executor

You can configure the executor that handles asynchronous MapStore operations as a regular Hazelcast Platform xref:computing:executor-service.adoc#configuring-executor-service[executor].
The offload executor's name is `hz:map-store-offloadable`.

[tabs] 
==== 

XML:: 
+ 
-- 
[source,xml]
----
<hazelcast>
    ...
    <executor-service name="hz:map-store-offloadable">
        <pool-size>1</pool-size>

>         <queue-capacity>10</queue-capacity>

Reviewer: better to have default settings here:
```
pool-size: 16
queue-capacity: 0
```

    </executor-service>
    ...
</hazelcast>
----
--

YAML::
+
[source,yaml]
----
hazelcast:
  ..
  executor-service:
    "hz:map-store-offloadable":
      pool-size: 1

>       queue-capacity: 10

Reviewer: same here, better to have default settings here:
```
pool-size: 16
queue-capacity: 0
```
